### PR TITLE
[SimplifyCFG] Use isWritableObject() API

### PR DESCRIPTION
**Pull request by nikic**
SimplifyCFG store speculation currently has some homegrown code to check for a writable object, handling the alloca special case only.

Switch it to use the generic isWritableObject() API, which means that we also support byval arguments, allocator return values, and writable arguments.

I've adjusted isWritableObject() to also check for the noalias attribute when handling writable. Otherwise, I don't think that we can generalize from at-entry writability. This was not relevant for previous uses of the function, because they'd already require noalias for other reasons anyway.

### CODE REVIEW

File: llvm/lib/Analysis/AliasAnalysis.cpp [hunk 1]
```diff
@@ -892,7 +892,10 @@ bool llvm::isWritableObject(const Value *Object,
     return true;
 
   if (auto *A = dyn_cast<Argument>(Object)) {
-    if (A->hasAttribute(Attribute::Writable)) {
+    // Also require noalias, otherwise writability at function entry cannot be
+    // generalized to writability at other program points, even if the pointer
+    // does not escape.
+    if (A->hasAttribute(Attribute::Writable) && A->hasNoAliasAttr()) {
```
Comment (goldsteinn) [inline llvm/lib/Analysis/AliasAnalysis.cpp:898]: Seems to the `noalias` check should be behind a bool. Think its liable to cause confusion that the API is assuming its callers intent when you can imagine scenarios where `noalias` is unnecessary.

File: llvm/lib/Transforms/Utils/SimplifyCFG.cpp [hunk 2]
```diff
@@ -3057,16 +3058,16 @@ static Value *isSafeToSpeculateStore(Instruction *I, BasicBlock *BrBB,
...
-        // also write to that same location.
-        // If the address of the local object never escapes the function, that
-        // means it's never concurrently read or written, hence moving the store
-        // from under the condition will not introduce a data race.
-        auto *AI = dyn_cast<AllocaInst>(getUnderlyingObject(StorePtr));
-        if (AI && !PointerMayBeCaptured(AI, false, true))
+        Value *Obj = getUnderlyingObject(StorePtr);
+        bool ExplicitlyDereferenceableOnly;
+        if (isWritableObject(Obj, ExplicitlyDereferenceableOnly) &&
+            !PointerMayBeCaptured(Obj, false, true) &&
```
Comment (goldsteinn) [inline llvm/lib/Transforms/Utils/SimplifyCFG.cpp:3064]: nit: While you at it:
`PointerMayBeCaptured(Obj, /*ReturnCaptures=*/false, /*StoreCaptures=*/true)`

File: llvm/lib/Analysis/AliasAnalysis.cpp (hunk 1 above)
Comment (nikic) [reply llvm/lib/Analysis/AliasAnalysis.cpp:898]: Part of the API contract for this function is:

> At later points, this is only the case if the pointer can not escape to a different thread.

Without noalias, this is not true.

Besides, it has "object" in the name, which in the context of AA refers to identified objects. A pointer argument is only an identified object if it's either byval or noalias.

**Review by dtcxzyw — APPROVED** [review]
LGTM.

**Pull request merged**